unittests/AST/ASTImporterGenericRedeclTest.cpp

```diff
@@ -99,6 +99,46 @@ struct RedeclChain : ASTImporterOptionSpecificTestBase {
   std::string getDefinition() { return TypeParam::Definition; }
   BindableMatcher<Decl> getPattern() const { return TypeParam().getPattern(); }
 
+  void CheckPreviousDecl(Decl *To0, Decl *To1) {
+    assert(To0 != To1);
+    assert(&To0->getASTContext() == &To1->getASTContext());
+
+    auto *ToTU = To0->getTranslationUnitDecl();
+
+    // Templates.
+    if (auto *ToT0 = dyn_cast<TemplateDecl>(To0)) {
+      auto *ToT1 = cast<TemplateDecl>(To1);
+      ASSERT_TRUE(ToT0->getTemplatedDecl());
+      ASSERT_TRUE(ToT1->getTemplatedDecl());
+      EXPECT_EQ(ToT1->getTemplatedDecl()->getPreviousDecl(),
+                ToT0->getTemplatedDecl());
+      return;
+    }
+
+    // Specializations.
+    if (auto *From0F = dyn_cast<FunctionDecl>(To0)) {
+      auto *To0F = cast<FunctionDecl>(To0);
+      if (From0F->getTemplatedKind() ==
+          FunctionDecl::TK_FunctionTemplateSpecialization) {
+        EXPECT_EQ(To0->getCanonicalDecl(), To1->getCanonicalDecl());
+        // There may be a hidden fwd spec decl before a spec decl.
+        // In that case the previous visible decl can be reached through that
+        // invisible one.
+        EXPECT_THAT(To0,
+                    testing::AnyOf(To1->getPreviousDecl(),
+                                   To1->getPreviousDecl()->getPreviousDecl()));
+        auto *TemplateD = FirstDeclMatcher<FunctionTemplateDecl>().match(
+            ToTU, functionTemplateDecl());
+        auto *FirstSpecD = *(TemplateD->spec_begin());
+        EXPECT_EQ(FirstSpecD->getCanonicalDecl(), To0F->getCanonicalDecl());
+        return;
+      }
+    }
+
+    // The rest: Classes, Functions, etc.
+    EXPECT_EQ(To1->getPreviousDecl(), To0);
```

Comment:
This check is not needed in the other cases?

Reply:
Good catch, indeed we must check this in case of templates too.